src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs

```diff
@@ -0,0 +1,1036 @@
+                newBitmap._nullBitmap = _nullBitmap == null ? null : (byte[])_nullBitmap.Clone();
+                newBitmap._columnsCount = _columnsCount;
+                return newBitmap;
+            }
+
+            internal void Clean()
+            {
+                _columnsCount = 0;
+                // no need to free _nullBitmap array - it is cached for the next row
+            }
+
+            /// <summary>
+            /// If this method returns true, the value is guaranteed to be null. This is not true vice versa:
+            /// if the bitmap value is false (if this method returns false), the value can be either null or non-null - no guarantee in this case.
+            /// To determine whether it is null or not, read it from the TDS (per NBCROW design spec, for IMAGE/TEXT/NTEXT columns server might send
+            /// bitmap = 0, when the actual value is null).
+            /// </summary>
+            internal bool IsGuaranteedNull(int columnOrdinal)
+            {
+                if (_columnsCount == 0)
+                {
+                    // not an NBC row
+                    return false;
+                }
+
+                Debug.Assert(columnOrdinal >= 0 && columnOrdinal < _columnsCount, "Invalid column ordinal");
+
+                byte testBit = (byte)(1 << (columnOrdinal & 0x7)); // columnOrdinal & 0x7 == columnOrdinal MOD 0x7
+                byte testByte = _nullBitmap[columnOrdinal >> 3];
+                return (testBit & testByte) != 0;
+            }
+        }
+
+
+        /////////////////////
+        // General methods //
+        /////////////////////
+
+        // If this object is part of a TdsParserSessionPool, then this *must* be called inside the pool's lock
+        internal void Activate(object owner)
+        {
+            Debug.Assert(_parser.MARSOn, "Can not activate a non-MARS connection");
+            Owner = owner; // must assign an owner for reclamation to work
+            int result = Interlocked.Increment(ref _activateCount);   // must have non-zero activation count for reclamation to work too.
+            Debug.Assert(result == 1, "invalid deactivate count");
+        }
+
+        // CancelRequest - use to cancel while writing a request to the server
+        //
+        // o none of the request might have been sent to the server, simply reset the buffer,
+        //   sending attention does not hurt
+        // o the request was partially written. Send an ignore header to the server. attention is
+        //   required if the server was waiting for data (e.g. insert bulk rows)
+        // o the request was completely written out and the server started to process the request.
+        //   attention is required to have the server stop processing.
+        //
+        internal void CancelRequest()
+        {
+            ResetBuffer();    // clear out unsent buffer
+            // If the first sqlbulkcopy timeout, _outputPacketNumber may not be 1,
```

Comment:
```suggestion
            // VSDD#903514, if the first sqlbulkcopy timeout, _outputPacketNumber may not be 1,
```
bring the code from netfx